mula/scheduler/storage/filters/casting.py

```diff
@@ -43,9 +43,9 @@ def cast_expression(expression: BinaryExpression, filter_: Filter) -> BinaryExpr
         # if the value can be decoded.
         try:
             decoded_value = json.loads(filter_.value)
-            if isinstance(decoded_value, dict):
-                # If it's a JSON object, return the expression as is. We don't
-                # need to cast it.
+            # If the string is a JSON object, return the expression as is.
+            # We don't need to cast it.
+            if isinstance(decoded_value, dict | list):
```

Comment:
Just to be sure; maybe you meant to use `isinstance(decoded_value, (dict, list))` instead?

As far as I know, `isinstance` doesn't know how to handle these union types

Reply:
Yeah I got two different linter suggestions. First the one you mentioned and second the one that is implemented. Both cases should be covered in the test. But I'll check this explicitly

Reply:
Ah I see. I would strongly recommend using a class type instead as that's what the signature suggests. I haven't seen unions being used in this context before. Not sure if this is the right example, but checking such types is also difficult, e.g.:
```
>>> type(2) is int
True
>>> type(2) is (int | str)
False
```

I'll leave it up to you

Reply:
> _Changed in version 3.10_: _classinfo_ can be a [Union Type](https://docs.python.org/3/library/stdtypes.html#types-union).

https://docs.python.org/3/library/functions.html#isinstance

I don't want to get involved, just out of curiosity because I don't know much about Python: how would classes be used to solve this? In TypeScript unions are a great zero-cost/compile-time abstraction. Are they unidiomatic in Python?

Reply:
In Python `is` returns whether it is exactly the same object in memoy, so that `type(2) is (int | str)` returns False doesn't really say anything. Since Python 3.10 you can indeed use the union type in `isinstance`:

```
In [1]: isinstance(["a"], dict|list)
Out[1]: True

In [2]: isinstance({"a": "b"}, dict|list)
Out[2]: True
```